src/peep/peep.c

```diff
@@ -3421,7 +3478,785 @@ static void peep_update_fixing(int steps, rct_peep* peep){
 		return;
 	}
 
-	RCT2_CALLPROC_X(RCT2_ADDRESS(0x006C0EB0, uint32)[peep->sub_state], steps, 0, 0, 0, (int)peep, (int)ride, 0);
+	switch (peep->sub_state) {
+		case 0:
+			loc_6C0EEC(steps, peep, ride);
+			break;
+
+		case 1:
+			loc_6C0FAA(steps, peep, ride);
+			break;
+
+		case 2:
+		case 3:
+		case 4:
+		case 5:
+			loc_6C100D(steps, peep, ride);
+			break;
+
+		case 6:
+			loc_6C10A5(steps, peep, ride);
+			break;
+
+		case 7:
+			loc_6C11CC(steps, peep, ride);
+			break;
+
+		case 8:
+			loc_6C121F(steps, peep, ride);
+			break;
+
+		case 9:
+			loc_6C133F(steps, peep, ride);
+			break;
+
+		case 10:
+			loc_6C13B4(steps, peep, ride);
+			break;
+
+		case 11:
+			loc_6C13F8(steps, peep, ride);
+			break;
+
+		case 12:
+			loc_6C14E5(steps, peep, ride);
+			break;
+
+		case 13:
+			loc_6C1559(steps, peep, ride);
+			break;
+
+		case 14:
+			loc_6C15EF(steps, peep, ride);
+			break;
+
+		default:
+			log_error("Invalid substate");
+			RCT2_CALLPROC_X(RCT2_ADDRESS(0x006C0EB0, uint32)[peep->sub_state], steps, 0, 0, 0, (int) peep, (int) ride, 0);
+	}
+
+}
+
+void loc_6C0EEC(int steps, rct_peep *peep, rct_ride *ride) {
+	// eax: steps
+	// esi: peep
+	// edi: ride
+
+	ride->mechanic_status = 3;
+	ride->window_invalidate_flags |= RIDE_INVALIDATE_RIDE_MAINTENANCE;
+
+	loc_6C1653(steps, peep, ride);
+}
+
+void loc_6C0F09(int steps, rct_peep *peep, rct_ride *ride) {
+	uint32 ebx = ride->broken_vehicle;
+	uint16 bp = ride->vehicles[peepDirection];
+
+	if (bp == 0xFFFF) {
+		loc_6C1653(steps, peep, ride);
+		return;
+	}
+
+	rct_sprite *ebp_sprite = &g_sprite_list[bp];
+	sint8 ah = ride->broken_car;
+
+	// loc_6C0F3E:
+	while (true) {
+		ah--;
+		if (ah < 0) {
+			break;
+		}
+
+		ebp_sprite = &g_sprite_list[ebp_sprite->vehicle.next_vehicle_on_train];
+	}
+
+	// loc_6C0F51:
+	while (true) {
+		if (ebp_sprite->vehicle.is_child == 0) {
+			break;
+		}
+
+		uint8 trackType = ebp_sprite->vehicle.track_type >> 2;
+		if (trackType == TRACK_ELEM_END_STATION) {
+			break;
+		}
+
+		if (trackType == TRACK_ELEM_BEGIN_STATION) {
+			break;
+		}
+
+		if (trackType == TRACK_ELEM_MIDDLE_STATION) {
+			break;
+		}
+
+		ebp_sprite = &g_sprite_list[ebp_sprite->vehicle.prev_vehicle_on_ride];
+	}
+
+	// loc_6C0F7A:
+	uint32 peepDirection = peep->var_78;
+
+	sint16 ax = word_981D6C[peepDirection].x;
+	sint16 cx = word_981D6C[peepDirection].y;
+
+	ax *= -12;
+	cx *= -12;
+
+	ax += ebp_sprite->vehicle.x;
+	cx += ebp_sprite->vehicle.y;
+
+	peep->destination_x = ax;
+	peep->destination_y = cx;
+	peep->destination_tolerence = 2;
+
+	loc_6C0FAA(steps, peep, ride);
+}
+
+static void loc_6C0FAA(int steps, rct_peep *peep, rct_ride *ride) {
+	// eax: steps
+	// esi: peep
+	// edi: ride
+
+	invalidate_sprite_2((rct_sprite *) peep);
+	int out = peep_update_action(); // TODO: arguments
+	if (out == 0) {
```

Comment:
This is the wrong way round. Peep_update_action return is basically the equivalent of the carry flag. So in this case it should goto `6c1653` when carry is zero.